src/ui/public/vis/editors/default/agg_params.js

```diff
@@ -40,7 +41,14 @@ uiModules
         $scope.$bind('agg', attr.agg);
         $scope.$bind('groupName', attr.groupName);
 
-        $scope.aggTypeOptions = aggTypes.byType[$scope.groupName];
+        const aggTypeSubscription = aggTypeFilters
```

Comment:
How important/valuable is it that we're exposing observables through the interface here? For example, I can also picture this functionality being supported by an interface that looked like this (while still using observables internally):

```js
const aggTypeSubscription = new AggTypeSubscription(aggTypes.byType[$scope.groupName], $scope.vis, $scope.agg);
aggTypeSubscription.onChange(aggTypes => $scope.aggTypeOptions = aggTypes);
```

This would encapsulate the complexity of observables and present me with an interface which appears and behaves like an event emitter.

Reply:
If we are anyway working with observables I think we are actually also want to expose this for a couple of reasons:

1) Otherwise we also need to build deregistration logic in that interface.
2) Observables are super powerful and allow tons of operators, so if the actual consumer of that API need to postprocess the answer in anyway or want to debounce it, or just listen for the first changes, etc. - there is already an Observable operator for that. I don't see a good reason why we would want to deliberately kill that flexibility and possibly force the user to wrap everything in an observable again to be able to achieve that operations.

Reply:
I have a couple thoughts in response, but keep in mind I'm coming to this with zero experience in observables, so I'm mostly asking questions to learn and not with any strongly held opinion. I would not consider my comments a blocker on this PR in any way.

> Observables are super powerful and allow tons of operators

100% agree, from what I've learned about them so far their power is apparent. But this is why I'm trying to learn more about why this power is necessary, since you don't want to use a tank when all you need is a bicycle. In this particular case, it doesn't seem like we're taking advantage of the observable. Are there concrete use cases for the `aggTypeFilters` module in which this power is necessary on the consumer's side, or examples of similar modules elsewhere that make the advantages more obvious?

> I don't see a good reason why we would want to deliberately kill that flexibility

Just to explain my thought process, I look for the abstractions we're creating in our code and try to figure out what kind of model can be inferred from them. So I poke and probe at abstractions that are complex and try to understand if that complexity is necessary. An example of where unnecessary complexity in abstractions created a confusing model would be our courier code. I was thinking that if it turns out that the observable is really just an implementation detail, then it might make sense to reduce complexity for the consumer by hiding it.

> Since Observables and RxJS tend to become a common pattern

No argument here. 😄 I agree that observables can be really useful and I'm sure there are situations where an abstraction that deals in observables makes sense. I'm just trying to understand why they're a core part of this abstraction's interface. I have very little experience with observables, so I'm really just trying to learn more about the situations in which they're most useful and understand how this is one of those situations.